File: app/subprojects/geologic_3d_engine/geologic_3d_engine/export/contract_integrity.py
"""Cross-language integrity envelope for the AutoCAD handoff contract."""

from __future__ import annotations

import base64
import hashlib
import json
import math
import unicodedata
from typing import Any

ENVELOPE_VERSION = "1.0"
CANONICALIZATION_VERSION = "GEO3D-JCS-LIKE-1.0"


def _normalize(value: Any) -> Any:
    if isinstance(value, str):
        return unicodedata.normalize("NFC", value)
    if isinstance(value, float):
        if not math.isfinite(value):
            raise ValueError("non-finite number is prohibited")
        return 0.0 if value == 0.0 else value
    if isinstance(value, dict):
        return {_normalize(str(key)): _normalize(item) for key, item in value.items()}
    if isinstance(value, (list, tuple)):
        return [_normalize(item) for item in value]
    if value is None or isinstance(value, (bool, int)):
        return value
    raise ValueError(f"unsupported contract value type: {type(value).__name__}")


def canonical_json_bytes(value: Any) -> bytes:
    return json.dumps(_normalize(value), ensure_ascii=False, sort_keys=True,
                      separators=(",", ":"), allow_nan=False).encode("utf-8")


def sha256_hex(payload: bytes) -> str:
    return hashlib.sha256(payload).hexdigest()


def build_integrity_envelope(payload: dict, validation: dict,
                             validator_version: str) -> dict:
    payload_bytes = canonical_json_bytes(payload)
    payload_digest = sha256_hex(payload_bytes)
    bound_validation = dict(validation)
    bound_validation.update({"payloadSha256": payload_digest,
                             "validatorVersion": validator_version})
    validation_bytes = canonical_json_bytes(bound_validation)
    return {"envelopeVersion": ENVELOPE_VERSION,
            "canonicalizationVersion": CANONICALIZATION_VERSION,
            "payloadEncoding": "base64-utf8-json",
            "payloadSha256": payload_digest,
            "payloadBase64": base64.b64encode(payload_bytes).decode("ascii"),
            "validationSha256": sha256_hex(validation_bytes),
            "validationBase64": base64.b64encode(validation_bytes).decode("ascii")}


def verify_integrity_envelope(envelope: dict) -> tuple[dict, dict]:
    if envelope.get("envelopeVersion") != ENVELOPE_VERSION:
        raise ValueError("unsupported envelopeVersion")
    if envelope.get("canonicalizationVersion") != CANONICALIZATION_VERSION:
        raise ValueError("unsupported canonicalizationVersion")
    if envelope.get("payloadEncoding") != "base64-utf8-json":
        raise ValueError("unsupported payloadEncoding")
    try:
        payload_bytes = base64.b64decode(envelope["payloadBase64"], validate=True)
        validation_bytes = base64.b64decode(envelope["validationBase64"], validate=True)
    except Exception as error:
        raise ValueError("invalid envelope base64") from error
    if sha256_hex(payload_bytes) != envelope.get("payloadSha256"):
        raise ValueError("payload digest mismatch")
    if sha256_hex(validation_bytes) != envelope.get("validationSha256"):
        raise ValueError("validation digest mismatch")
    payload = json.loads(payload_bytes.decode("utf-8"))
    validation = json.loads(validation_bytes.decode("utf-8"))
    if validation.get("payloadSha256") != envelope["payloadSha256"]:
        raise ValueError("validation is not bound to payload")
    if not validation.get("passed", False):
        raise ValueError("validation did not pass")
    if canonical_json_bytes(payload) != payload_bytes:
        raise ValueError("payload violates canonicalization contract")
    if canonical_json_bytes(validation) != validation_bytes:
        raise ValueError("validation violates canonicalization contract")
    return payload, validation

File: app/tools/AuthoritativeGeoDwg/ContractIntegrity.cs
using System.Security.Cryptography;
using System.Text;
using System.Text.Json;

namespace AuthoritativeGeoDwg;

/// <summary>AutoCAD-independent, transaction-free contract preflight.</summary>
public static class ContractIntegrity
{
    public const string EnvelopeVersion = "1.0";
    public const string CanonicalizationVersion = "GEO3D-JCS-LIKE-1.0";

    public static JsonDocument VerifyAndParse(JsonElement documentRoot)
    {
        var envelope = documentRoot.TryGetProperty("geologicContractEnvelope", out var nested)
            ? nested : documentRoot;
        RequireString(envelope, "envelopeVersion", EnvelopeVersion);
        RequireString(envelope, "canonicalizationVersion", CanonicalizationVersion);
        RequireString(envelope, "payloadEncoding", "base64-utf8-json");

        var payloadBytes = Decode(envelope, "payloadBase64");
        var validationBytes = Decode(envelope, "validationBase64");
        var payloadDigest = RequireString(envelope, "payloadSha256");
        var validationDigest = RequireString(envelope, "validationSha256");
        if (!FixedEquals(Sha256(payloadBytes), payloadDigest))
            throw new InvalidDataException("payload digest mismatch");
        if (!FixedEquals(Sha256(validationBytes), validationDigest))
            throw new InvalidDataException("validation digest mismatch");

        using var validation = JsonDocument.Parse(validationBytes);
        var validationRoot = validation.RootElement;
        if (!validationRoot.GetProperty("passed").GetBoolean())
            throw new InvalidDataException("validation did not pass");
        if (validationRoot.GetProperty("payloadSha256").GetString() != payloadDigest)
            throw new InvalidDataException("validation is not bound to payload");

        var payload = JsonDocument.Parse(payloadBytes);
        ValidateFinite(payload.RootElement);
        var root = payload.RootElement;
        if (!root.TryGetProperty("lineageManifest", out var lineage) ||
            !lineage.TryGetProperty("sources", out var sources) || sources.GetArrayLength() == 0)
        {
            payload.Dispose();
            throw new InvalidDataException("source-linked lineageManifest is required");
        }
        ValidatePayload(root, validationRoot);
        return payload;
    }

    private static byte[] Decode(JsonElement root, string name)
    {
        try { return Convert.FromBase64String(root.GetProperty(name).GetString()!); }
        catch (Exception error) { throw new InvalidDataException($"invalid {name}", error); }
    }

    private static string RequireString(JsonElement root, string name, string? expected = null)
    {
        if (!root.TryGetProperty(name, out var value) || value.ValueKind != JsonValueKind.String)
            throw new InvalidDataException($"missing {name}");
        var text = value.GetString()!;
        if (expected is not null && text != expected)
            throw new InvalidDataException($"unsupported {name}");
        return text;
    }

    private static string Sha256(byte[] bytes) => Convert.ToHexString(SHA256.HashData(bytes)).ToLowerInvariant();

    private static bool FixedEquals(string actual, string expected)
    {
        if (actual.Length != expected.Length) return false;
        return CryptographicOperations.FixedTimeEquals(Encoding.ASCII.GetBytes(actual), Encoding.ASCII.GetBytes(expected));
    }

    private static void ValidateFinite(JsonElement value)
    {
        if (value.ValueKind == JsonValueKind.Number && !double.IsFinite(value.GetDouble()))
            throw new InvalidDataException("non-finite number is prohibited");
        if (value.ValueKind == JsonValueKind.Array)
            foreach (var item in value.EnumerateArray()) ValidateFinite(item);
        if (value.ValueKind == JsonValueKind.Object)
            foreach (var item in value.EnumerateObject()) ValidateFinite(item.Value);
    }

    private static void ValidatePayload(JsonElement root, JsonElement validation)
    {
        RequireString(root, "dwgVersion", "AC1032");
        if (!root.TryGetProperty("notForDesign", out var disclosure) || !disclosure.GetBoolean())
            throw new InvalidDataException("synthetic not-for-design disclosure is required");
        if (!root.TryGetProperty("polygons", out var polygons) || polygons.ValueKind != JsonValueKind.Array || polygons.GetArrayLength() == 0)
            throw new InvalidDataException("non-empty polygons are required");
        if (validation.GetProperty("polygonCount").GetInt32() != polygons.GetArrayLength())
            throw new InvalidDataException("validation polygon count mismatch");
        var ids = new HashSet<string>(StringComparer.Ordinal);
        foreach (var polygon in polygons.EnumerateArray())
        {
            var id = RequireString(polygon, "polygonId");
            if (!ids.Add(id)) throw new InvalidDataException("duplicate polygonId");
            RequireString(polygon, "unitId");
            RequireString(polygon, "boundaryLayer");
            RequireString(polygon, "hatchLayer");
            RequireString(polygon, "hatchPattern", "SOLID");
            if (!polygon.GetProperty("closed").GetBoolean() || !polygon.GetProperty("associative").GetBoolean())
                throw new InvalidDataException("closed associative polygon is required");
            var rgb = polygon.GetProperty("trueColorRGB");
            if (rgb.GetArrayLength() != 3 || rgb.EnumerateArray().Any(v => v.GetInt32() is < 0 or > 255))
                throw new InvalidDataException("invalid trueColorRGB");
            var vertices = polygon.GetProperty("vertices");
            if (vertices.GetArrayLength() < 3 || vertices.EnumerateArray().Any(v => v.GetArrayLength() != 2))
                throw new InvalidDataException("at least three UV vertices are required");
        }
    }
}
